src/Microsoft.DotNet.Darc/src/DarcLib/Models/Darc/DependencyFlowGraph.cs

```diff
@@ -127,20 +127,219 @@ public void RemoveEdge(DependencyFlowEdge edge)
+            List<DependencyFlowEdge> toRemove = new List<DependencyFlowEdge>();
+            foreach (var edge in Edges)
+            {
+                if (dominators[edge.To].Contains(edge.From))
+                {
+                    edge.BackEdge = true;
+                }
+            }
+
+            // Remove the start node we created and links to it
+            foreach (var edge in startNode.IncomingEdges)
+            {
+                RemoveEdge(edge);
+            }
+
+            RemoveNode(startNode);
+        }
+
+        /// <summary>
+        ///     Calculate the longest build path time from each node in the flow graph
+        /// </summary>
+        /// <remarks>
+        ///     Starting with the set of root node (nodes with no outgoing edges), compute the
+        ///     longest build path time using a breadth first search
+        /// </remarks>
+        public void CalculateLongestBuildPaths()
+        {
+            List<DependencyFlowNode> roots = Nodes.Where(n => n.OutgoingEdges.Count == 0).ToList();
+            Dictionary<DependencyFlowNode, List<DependencyFlowNode>> visitedNodes = new Dictionary<DependencyFlowNode, List<DependencyFlowNode>>();
+
+            Queue<DependencyFlowNode> nodesToVisit = new Queue<DependencyFlowNode>();
+
+            foreach (var root in roots)
+            {
+                nodesToVisit.Enqueue(root);
+
+                while (nodesToVisit.Count > 0)
+                {
+                    DependencyFlowNode node = nodesToVisit.Dequeue();
+                    if (visitedNodes.ContainsKey(node))
+                    {
+                        visitedNodes[node].Add(node);
+                    }
+                    else
+                    {
+                        visitedNodes.Add(node, new List<DependencyFlowNode>(){node});
+                    }
+
+                    foreach (var edge in node.IncomingEdges)
+                    {
+                        DependencyFlowNode child = edge.From;
+
+                        if (!visitedNodes[node].Contains(child) && !nodesToVisit.Contains(child))
+                        {
+                            if (visitedNodes.ContainsKey(child))
+                            {
+                                visitedNodes[child].AddRange(visitedNodes[node]);
+                            }
+                            else
+                            {
+                                visitedNodes.Add(child, new List<DependencyFlowNode>(visitedNodes[node]));
+                            }
+                            
+                            nodesToVisit.Enqueue(child);
+                        }
+                    }
+
+                    node.CalculateLongestPathTime();
+                }
+            }
+        }
+
+        /// <summary>
+        ///     Determine and mark the absolute longest build path in the flow graph, based on the Best Case time.
+        /// </summary>
+        public void MarkLongestBuildPath()
```

Comment:
Should these really be public? They seem like private implementation details of how some of the other methods operate.

Reply:
These methods are all similar to PruneGraph and Build, where they are called by the darc getDependencyFlowGraphOperation.